amy/extrequests/forms.py

```diff
@@ -1407,6 +1407,79 @@ class Meta:
             "state": forms.RadioSelect(),
         }
 
+    def __init__(self, *args, **kwargs):
+        # request is required for ENFORCE_MEMBER_CODES flag
+        self.request_http = kwargs.pop("request", None)
+        super().__init__(*args, **kwargs)
+
+    def clean(self):
+        super().clean()
+
+        errors = self.validate_member_code(request=self.request_http)
+        if errors:
+            raise ValidationError(errors)
+
+    @feature_flag_enabled("ENFORCE_MEMBER_CODES")
+    def validate_member_code(
+        self, request: HttpRequest
+    ) -> None | dict[str, ValidationError]:
+        errors = dict()
+        member_code = self.cleaned_data.get("member_code", "")
+        member_code_override = self.cleaned_data.get("member_code_override", False)
+
+        if not member_code:
+            return None
+
+        member_code_valid, error_detail = self.member_code_valid(member_code)
+        if member_code_valid and member_code_override:
+            # case where a user corrects their code but ticks the box anyway
+            # checkbox doesn't need to be ticked, so correct it quietly and continue
+            self.cleaned_data["member_code_override"] = False
+        elif not member_code_valid:
+            if not member_code_override:
+                # user must either correct the code or tick the override
+                error_msg = (
+                    "This code is invalid: "
+                    f"{error_detail} "
+                    "Tick the checkbox below to ignore this message."
+                )
+                errors["member_code"] = ValidationError(error_msg)
+
+        return errors
+
+    def member_code_valid(self, code: str) -> tuple[bool, str]:
+        """Returns True if the code matches an active Membership with available seats,
+        including a grace period of 90 days before and after the Membership dates.
+        If there is no match, returns False and a detailed error.
+        """
+        try:
+            # find relevant membership - may raise Membership.DoesNotExist
+            membership = Membership.objects.get(registration_code=code)
+        except Membership.DoesNotExist:
+            return False, f'No membership found for code "{code}".'
```

Comment:
You could use custom exceptions instead of returning a flag and text.
For example:

```python

class MemberCodeValidationError(ValueError):  # or inherit from ValidationError
  pass

...

    def member_code_valid(self, code: str) -> bool:
        """Returns True if the code matches an active Membership with available seats,
        including a grace period of 90 days before and after the Membership dates.
        If there is no match, throws an exception with detailed error.
        """
        try:
            # find relevant membership - may raise Membership.DoesNotExist
            membership = Membership.objects.get(registration_code=code)
        except Membership.DoesNotExist as exc:
            raise MemberCodeValidationError(f'No membership found for code "{code}".') from exc

        ...
        return True

```